src/main/scala/beam/physsim/PickUpDropOffCollector.scala

```diff
@@ -0,0 +1,270 @@
+    }
+
+    val filteredVehiclePickUpsDropOffs = vehicleToPickUpsDropOffs
+      .filter { case (vehicleId, _) => !vehiclesToIgnore.contains(vehicleId) }
+
+    personsPickUps.foreach { personPickUp =>
+      filteredVehiclePickUpsDropOffs.get(personPickUp.vehicleId) match {
+        case Some(vehiclePickUpsDropOffs) =>
+          vehiclePickUpsDropOffs.pickUpTimeToLink.getClosestValue(personPickUp.time) match {
+            case Some(link) => addLinkPickUp(link, personPickUp.time)
+            case None =>
+              logger.error(
+                s"Missing vehicle pick up point from PathTraversals for vehicle ${personPickUp.vehicleId} at time ${personPickUp.time}. All data: $vehiclePickUpsDropOffs"
+              )
+          }
+        case None =>
+      }
+    }
+
+    personsDropOffs.foreach { personDropOff =>
+      filteredVehiclePickUpsDropOffs.get(personDropOff.vehicleId) match {
+        case Some(vehiclePickUpsDropOffs) =>
+          vehiclePickUpsDropOffs.dropOffTimeToLink.getClosestValue(personDropOff.time) match {
+            case Some(link) => addLinkDropOff(link, personDropOff.time)
+            case None =>
+              logger.error(
+                s"Missing vehicle drop off point from PathTraversals for vehicle ${personDropOff.vehicleId} at time ${personDropOff.time}. All data: $vehiclePickUpsDropOffs"
+              )
+          }
+        case None =>
+      }
+    }
+
+    linkToPickUpsDropOffs
+  }
+}
+
+case class LinkPickUpsDropOffs(
+  linkId: Int,
+  timeToPickUps: TimeToValueCollection = new TimeToValueCollection(),
+  timeToDropOffs: TimeToValueCollection = new TimeToValueCollection()
+) {
+  override def toString: String = s"link $linkId has pickUps: $timeToPickUps dropOffs: $timeToDropOffs"
+}
+
+object PersonPickUp {
+
+  def apply(pev: PersonEntersVehicleEvent): PersonPickUp =
+    new PersonPickUp(pev.getTime, pev.getVehicleId)
+}
+case class PersonPickUp(time: Double, vehicleId: Id[Vehicle])
+
+object PersonDropOff {
+
+  def apply(plv: PersonLeavesVehicleEvent): PersonDropOff =
+    new PersonDropOff(plv.getTime, plv.getVehicleId)
+}
+case class PersonDropOff(time: Double, vehicleId: Id[Vehicle])
+
+case class VehiclePickUpsDropOffs(
+  pickUpTimeToLink: TimeToValueCollection = new TimeToValueCollection(),
+  dropOffTimeToLink: TimeToValueCollection = new TimeToValueCollection()
+) {
+
+  def addPickUpDropOff(pte: PathTraversalEvent): Unit = {
+    pickUpTimeToLink.addTimeToValue(pte.departureTime, pte.linkIds.head)
+    dropOffTimeToLink.addTimeToValue(pte.arrivalTime, pte.linkIds.last)
+  }
+
+  override def toString: String = s"PickUps: $pickUpTimeToLink DropOffs: $dropOffTimeToLink"
+}
+
+class TimeToValueCollection() {
+  val times: mutable.ListBuffer[Double] = mutable.ListBuffer.empty
+  val values: mutable.ListBuffer[Int] = mutable.ListBuffer.empty
+
+  // I assume time will be added already in sorted order
+  def addTimeToValue(time: Double, value: Int): Unit = {
```

Comment:
I'm not sure this can be tested for...can we really guarantee this over time? Should we use a sorted list to guarantee?

Reply:
ok, I wrote the code in case time is not added in sorted order